{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tri-allelic SNPs Illumina (re-run)\n",
    "\n",
    "* New libraries\n",
    "* Original fastq: 9948 and 9947 are switched!\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg exists\n",
      "Directory /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data exists\n",
      "Directory /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/mapped_data exists\n"
     ]
    }
   ],
   "source": [
    "# Init\n",
    "#\n",
    "import os, glob\n",
    "\n",
    "projectDir    = '/media/genomics/nanopore/projects/tri-allelic_SNPs/20180219_illumina_analysis_yg'\n",
    "rawDataDir    = '/media/genomics/nanopore/run_data/20180219_Illumina_tri-allelic/PCRfree_Senne-65584528'\n",
    "fastQDir      = os.path.join(projectDir, 'raw_data')\n",
    "mappedDataDir = os.path.join(projectDir, 'mapped_data')\n",
    "snpRefFile    = os.path.join(projectDir, 'triallelic_snp_regions.fasta')\n",
    "sampleList    = ['2800', '9947', '9948', 'G59', 'G62']\n",
    "\n",
    "bwa           = '/opt/tools/bwa-0.7.15'\n",
    "samtools      = '/opt/tools/samtools-1.3.1'\n",
    "bcftools      = '/opt/tools/bcftools-1.3.1'\n",
    "\n",
    "# Create dirs\n",
    "for d in [projectDir, fastQDir, mappedDataDir]:\n",
    "  if not os.path.exists(d):\n",
    "    print('Creating directory {}'.format(d))\n",
    "    os.makedirs(d)\n",
    "  else:\n",
    "    print('Directory {} exists'.format(d))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2800_S1_L001_R1_001.fastq -> 2800_R1.fastq\n",
      "2800_S1_L001_R2_001.fastq -> 2800_R2.fastq\n",
      "9947_S3_L001_R1_001.fastq -> 9948_R1.fastq\n",
      "9947_S3_L001_R2_001.fastq -> 9948_R2.fastq\n",
      "9948_S2_L001_R1_001.fastq -> 9947_R1.fastq\n",
      "9948_S2_L001_R2_001.fastq -> 9947_R2.fastq\n",
      "G59_S5_L001_R1_001.fastq -> G59_R1.fastq\n",
      "G59_S5_L001_R2_001.fastq -> G59_R2.fastq\n",
      "G62_S4_L001_R1_001.fastq -> G62_R1.fastq\n",
      "G62_S4_L001_R2_001.fastq -> G62_R2.fastq\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# Copy and rename fastq files (handle swithed samples)\n",
    "#\n",
    "fastqNameMap = {\n",
    "  '2800_S1_L001_R1_001.fastq': '2800_R1.fastq',\n",
    "  '2800_S1_L001_R2_001.fastq': '2800_R2.fastq',\n",
    "  '9947_S3_L001_R1_001.fastq': '9948_R1.fastq',\n",
    "  '9947_S3_L001_R2_001.fastq': '9948_R2.fastq',\n",
    "  '9948_S2_L001_R1_001.fastq': '9947_R1.fastq',\n",
    "  '9948_S2_L001_R2_001.fastq': '9947_R2.fastq',\n",
    "  'G59_S5_L001_R1_001.fastq':  'G59_R1.fastq',\n",
    "  'G59_S5_L001_R2_001.fastq':  'G59_R2.fastq',\n",
    "  'G62_S4_L001_R1_001.fastq':  'G62_R1.fastq',\n",
    "  'G62_S4_L001_R2_001.fastq':  'G62_R2.fastq'\n",
    "}\n",
    "\n",
    "for fastqOld, fastqNew in fastqNameMap.items():\n",
    "  print('{} -> {}'.format(fastqOld, fastqNew))\n",
    "  ! cp {os.path.join(rawDataDir, fastqOld)} {os.path.join(fastQDir, fastqNew)}\n",
    "  \n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.00 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.00 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.00 sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 index /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta\n",
      "[main] Real time: 0.026 sec; CPU: 0.009 sec\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# Index reference\n",
    "#\n",
    "! {bwa} index {snpRefFile}\n",
    "\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 381286 sequences (57574186 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 29066, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (50, 50, 50)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (50, 50)\n",
      "[M::mem_pestat] mean and std.dev: (50.00, 0.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (50, 50)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 381286 reads in 22.158 CPU sec, 2.437 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 mem -t 10 /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/2800_R1.fastq /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/2800_R2.fastq\n",
      "[main] Real time: 11.193 sec; CPU: 23.277 sec\n",
      "Note: Neither --ploidy nor --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 453520 sequences (68481520 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 25229, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (49, 50, 50)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (47, 52)\n",
      "[M::mem_pestat] mean and std.dev: (49.63, 0.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (46, 53)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 453520 reads in 27.724 CPU sec, 3.077 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 mem -t 10 /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/9947_R1.fastq /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/9947_R2.fastq\n",
      "[main] Real time: 12.178 sec; CPU: 28.770 sec\n",
      "Note: Neither --ploidy nor --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 417512 sequences (63044312 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 13091, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (49, 50, 50)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (47, 52)\n",
      "[M::mem_pestat] mean and std.dev: (49.60, 0.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (46, 53)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 417512 reads in 24.344 CPU sec, 2.756 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 mem -t 10 /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/9948_R1.fastq /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/9948_R2.fastq\n",
      "[main] Real time: 11.371 sec; CPU: 25.582 sec\n",
      "Note: Neither --ploidy nor --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 566208 sequences (85497408 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 62841, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (49, 50, 50)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (47, 52)\n",
      "[M::mem_pestat] mean and std.dev: (49.65, 0.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (46, 53)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 566208 reads in 33.393 CPU sec, 3.648 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 mem -t 10 /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/G59_R1.fastq /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/G59_R2.fastq\n",
      "[main] Real time: 15.472 sec; CPU: 34.706 sec\n",
      "Note: Neither --ploidy nor --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 430086 sequences (64942986 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 63906, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (50, 50, 50)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (50, 50)\n",
      "[M::mem_pestat] mean and std.dev: (50.00, 0.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (50, 50)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 430086 reads in 28.175 CPU sec, 3.062 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: /opt/tools/bwa-0.7.15 mem -t 10 /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/triallelic_snp_regions.fasta /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/G62_R1.fastq /media/genomics/nanopore/tri-allelic_SNPS/20180219_illumina_analysis_yg/raw_data/G62_R2.fastq\n",
      "[main] Real time: 12.871 sec; CPU: 29.575 sec\n",
      "Note: Neither --ploidy nor --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# Map reads to reference SNP region sequences (51 nt), call variants\n",
    "#\n",
    "for sampleName in sorted(sampleList):\n",
    "  # Map amplicons\n",
    "  read1File = os.path.join(fastQDir, '{}_R1.fastq'.format(sampleName))\n",
    "  read2File = os.path.join(fastQDir, '{}_R2.fastq'.format(sampleName))\n",
    "  bamFile   = os.path.join(mappedDataDir, '{}_direct_mapping.bam'.format(sampleName))\n",
    "  vcfFile   = bamFile.replace('.bam', '.vcf')\n",
    "  \n",
    "  ! {bwa} mem -t 10 {snpRefFile} {read1File}  {read2File} | {samtools} view -Sb - | {samtools} sort -o {bamFile} -\n",
    "  ! {samtools} index {bamFile}\n",
    "  \n",
    "  # Variant calling\n",
    "  !{samtools} mpileup -d 100000 -Buf {snpRefFile} -t AD {bamFile} | {bcftools} call -V indels -m - > {vcfFile}\n",
    "  \n",
    "\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def genotype(vcfFile, snpPos):\n",
    "  \"\"\"\n",
    "  Extract genotype data at given position from a given vcf file.\n",
    "  \"\"\"\n",
    "  genotypeData = {}\n",
    "  \n",
    "  with open (vcfFile, 'rt') as inFile:\n",
    "    for line in inFile:\n",
    "      if line.startswith('#'): continue\n",
    "      \n",
    "      snpName, pos, id, ref, alt, qual, filt, info, form, formValues = line.rstrip().split()\n",
    "      \n",
    "      if int(pos) == snpPos:\n",
    "        k        = form.split(':')\n",
    "        v        = formValues.split(':')         \n",
    "        formData = {}\n",
    "        \n",
    "        for i in range(len(k)):\n",
    "          formData[k[i]] = v[i]\n",
    "          \n",
    "        infoData = {}\n",
    "        for t in info.split(';'):\n",
    "          k,v = t.split('=')\n",
    "          infoData[k] = v\n",
    "          \n",
    "        alleles = ref\n",
    "        if alt != '.':\n",
    "          alleles += ''.join(alt.split(','))\n",
    "          \n",
    "        depths                = [int(d) for d in formData['AD'].split(',')]\n",
    "        gt                    = [alleles[int(i)] for i in formData['GT'].split('/') ]\n",
    "        genotypeData[snpName] = {'pos': pos, 'alleles': {'A': 0, 'G': 0, 'C': 0, 'T':0}, 'genotype': '/'.join(gt), 'depth': int(infoData['DP'])}\n",
    "        \n",
    "        for i in range(len(alleles)):\n",
    "          genotypeData[snpName]['alleles'][alleles[i]] = depths[i]\n",
    "        \n",
    "  return genotypeData\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP's for sample 2800\n",
      "  rs1008686         A:0      C:0      G:0      T:1928   A:  0%  C:  0%  G:  0%  T:100%     1928/3876  reads    TT\n",
      "  rs1112534         A:0      C:1347   G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%     1347/2700  reads    CC\n",
      "  rs17287498        A:253    C:1      G:318    T:0      A: 44%  C:  0%  G: 56%  T:  0%      572/1144  reads    AG\n",
      "  rs2032582         A:1879   C:0      G:8      T:1844   A: 50%  C:  0%  G:  0%  T: 49%     3731/7471  reads    AT\n",
      "  rs2069945         A:0      C:0      G:423    T:0      A:  0%  C:  0%  G:100%  T:  0%      423/848   reads    GG\n",
      "  rs2307223         A:2240   C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%     2240/4492  reads    AA\n",
      "  rs2853525         A:0      C:415    G:0      T:367    A:  0%  C: 53%  G:  0%  T: 47%      782/1564  reads    CT\n",
      "  rs3091244         A:0      C:148    G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%      148/298   reads    CC\n",
      "  rs34741930        A:0      C:2      G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%        2/4     reads    CC\n",
      "  rs35528968        A:880    C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%      880/1764  reads    AA\n",
      "  rs356167          A:976    C:0      G:1275   T:0      A: 43%  C:  0%  G: 57%  T:  0%     2251/4542  reads    AG\n",
      "  rs433342          A:0      C:0      G:27     T:0      A:  0%  C:  0%  G:100%  T:  0%       27/54    reads    GG\n",
      "  rs5030240         A:0      C:0      G:1994   T:1899   A:  0%  C:  0%  G: 51%  T: 49%     3893/7802  reads    GT\n",
      "  rs727241          A:0      C:0      G:0      T:1092   A:  0%  C:  0%  G:  0%  T:100%     1092/2194  reads    TT\n",
      "  rs9274701_alt     A:0      C:0      G:1858   T:0      A:  0%  C:  0%  G:100%  T:  0%     1858/3742  reads    GG\n",
      "  rs9329104         A:308    C:0      G:307    T:0      A: 50%  C:  0%  G: 50%  T:  0%      615/1234  reads    AG\n",
      "\n",
      "SNP's for sample 9947\n",
      "  rs1008686         A:0      C:0      G:0      T:2068   A:  0%  C:  0%  G:  0%  T:100%     2068/4144  reads    TT\n",
      "  rs1112534         A:0      C:451    G:0      T:388    A:  0%  C: 54%  G:  0%  T: 46%      839/1678  reads    CT\n",
      "  rs17287498        A:0      C:0      G:494    T:0      A:  0%  C:  0%  G:100%  T:  0%      494/988   reads    GG\n",
      "  rs2032582         A:0      C:0      G:1904   T:1963   A:  0%  C:  0%  G: 49%  T: 51%     3867/7758  reads    GT\n",
      "  rs2069945         A:0      C:383    G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%      383/770   reads    CC\n",
      "  rs2307223         A:651    C:0      G:0      T:640    A: 50%  C:  0%  G:  0%  T: 50%     1291/2590  reads    AT\n",
      "  rs2853525         A:0      C:341    G:0      T:294    A:  0%  C: 54%  G:  0%  T: 46%      635/1269  reads    CT\n",
      "  rs3091244         A:0      C:64     G:0      T:56     A:  0%  C: 53%  G:  0%  T: 47%      120/240   reads    CT\n",
      "  rs34741930        A:1      C:1867   G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%     1868/3737  reads    CC\n",
      "  rs35528968        A:811    C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%      811/1623  reads    AA\n",
      "  rs356167          A:0      C:943    G:1159   T:0      A:  0%  C: 45%  G: 55%  T:  0%     2102/4227  reads    CG\n",
      "  rs433342          A:0      C:2547   G:2361   T:0      A:  0%  C: 52%  G: 48%  T:  0%     4908/9827  reads    CG\n",
      "  rs5030240         A:0      C:877    G:1041   T:0      A:  0%  C: 46%  G: 54%  T:  0%     1918/3842  reads    CG\n",
      "  rs727241          A:0      C:0      G:0      T:784    A:  0%  C:  0%  G:  0%  T:100%      784/1578  reads    TT\n",
      "  rs9274701_alt     A:0      C:1026   G:946    T:0      A:  0%  C: 52%  G: 48%  T:  0%     1972/3948  reads    CG\n",
      "  rs9329104         A:209    C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%      209/424   reads    AA\n",
      "\n",
      "SNP's for sample 9948\n",
      "  rs1008686         A:0      C:0      G:0      T:1951   A:  0%  C:  0%  G:  0%  T:100%     1951/3915  reads    TT\n",
      "  rs1112534         A:0      C:217    G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%      217/434   reads    CC\n",
      "  rs17287498        A:0      C:0      G:249    T:0      A:  0%  C:  0%  G:100%  T:  0%      249/502   reads    GG\n",
      "  rs2032582         A:0      C:0      G:2415   T:0      A:  0%  C:  0%  G:100%  T:  0%     2415/4857  reads    GG\n",
      "  rs2069945         A:0      C:1      G:40     T:0      A:  0%  C:  2%  G: 98%  T:  0%       41/82    reads    GG\n",
      "  rs2307223         A:236    C:0      G:0      T:267    A: 47%  C:  0%  G:  0%  T: 53%      503/1010  reads    AT\n",
      "  rs2853525         A:0      C:454    G:0      T:6      A:  0%  C: 99%  G:  0%  T:  1%      460/918   reads    CC\n",
      "  rs3091244         A:33     C:0      G:0      T:19     A: 63%  C:  0%  G:  0%  T: 37%       52/106   reads    AT\n",
      "  rs34741930        A:0      C:885    G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%      885/1771  reads    CC\n",
      "  rs35528968        A:212    C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%      212/424   reads    AA\n",
      "  rs356167          A:290    C:0      G:419    T:0      A: 41%  C:  0%  G: 59%  T:  0%      709/1428  reads    AG\n",
      "  rs433342          A:0      C:2      G:2982   T:0      A:  0%  C:  0%  G:100%  T:  0%     2984/5978  reads    GG\n",
      "  rs5030240         A:0      C:0      G:762    T:0      A:  0%  C:  0%  G:100%  T:  0%      762/1550  reads    GG\n",
      "  rs727241          A:0      C:0      G:0      T:172    A:  0%  C:  0%  G:  0%  T:100%      172/346   reads    TT\n",
      "  rs9274701_alt     A:0      C:467    G:270    T:0      A:  0%  C: 63%  G: 37%  T:  0%      737/1476  reads    CG\n",
      "  rs9329104         A:51     C:0      G:46     T:0      A: 53%  C:  0%  G: 47%  T:  0%       97/194   reads    AG\n",
      "\n",
      "SNP's for sample G59\n",
      "  rs1008686         A:2715   C:0      G:0      T:1694   A: 62%  C:  0%  G:  0%  T: 38%     4409/8826  reads    AT\n",
      "  rs1112534         A:0      C:1854   G:28     T:0      A:  0%  C: 99%  G:  1%  T:  0%     1882/3770  reads    CC\n",
      "  rs17287498        A:0      C:2      G:923    T:1172   A:  0%  C:  0%  G: 44%  T: 56%     2097/4193  reads    GT\n",
      "  rs2032582         A:3635   C:0      G:13     T:2690   A: 57%  C:  0%  G:  0%  T: 42%     6338/12693 reads    AT\n",
      "  rs2069945         A:698    C:898    G:0      T:0      A: 44%  C: 56%  G:  0%  T:  0%     1596/3340  reads    AC\n",
      "  rs2307223         A:1408   C:0      G:1504   T:0      A: 48%  C:  0%  G: 52%  T:  0%     2912/5824  reads    AG\n",
      "  rs2853525         A:0      C:0      G:0      T:1552   A:  0%  C:  0%  G:  0%  T:100%     1552/3108  reads    TT\n",
      "  rs3091244         A:0      C:264    G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%      264/529   reads    CC\n",
      "  rs34741930        A:0      C:6328   G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%     6328/12660 reads    CC\n",
      "  rs35528968        A:2619   C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%     2619/5257  reads    AA\n",
      "  rs356167          A:0      C:4189   G:4420   T:0      A:  0%  C: 49%  G: 51%  T:  0%     8609/17364 reads    CG\n",
      "  rs433342          A:3258   C:4      G:4535   T:0      A: 42%  C:  0%  G: 58%  T:  0%     7797/15583 reads    AG\n",
      "  rs5030240         A:0      C:0      G:3998   T:2708   A:  0%  C:  0%  G: 60%  T: 40%     6706/13459 reads    GT\n",
      "  rs727241          A:0      C:1389   G:0      T:1244   A:  0%  C: 53%  G:  0%  T: 47%     2633/5270  reads    CT\n",
      "  rs9274701_alt     A:0      C:3952   G:3      T:0      A:  0%  C:100%  G:  0%  T:  0%     3955/7919  reads    CC\n",
      "  rs9329104         A:1693   C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%     1693/3402  reads    AA\n",
      "\n",
      "SNP's for sample G62\n",
      "  rs1008686         A:4439   C:0      G:0      T:6      A:100%  C:  0%  G:  0%  T:  0%     4445/8898  reads    AA\n",
      "  rs1112534         A:0      C:0      G:1      T:2251   A:  0%  C:  0%  G:  0%  T:100%     2252/4518  reads    TT\n",
      "  rs17287498        A:0      C:0      G:1875   T:0      A:  0%  C:  0%  G:100%  T:  0%     1875/3755  reads    GG\n",
      "  rs2032582         A:0      C:0      G:7368   T:0      A:  0%  C:  0%  G:100%  T:  0%     7368/14780 reads    GG\n",
      "  rs2069945         A:0      C:1251   G:712    T:0      A:  0%  C: 64%  G: 36%  T:  0%     1963/3930  reads    CG\n",
      "  rs2307223         A:2932   C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%     2932/5884  reads    AA\n",
      "  rs2853525         A:0      C:0      G:0      T:1732   A:  0%  C:  0%  G:  0%  T:100%     1732/3562  reads    TT\n",
      "  rs3091244         A:0      C:256    G:1      T:0      A:  0%  C:100%  G:  0%  T:  0%      257/514   reads    CC\n",
      "  rs34741930        A:0      C:13     G:0      T:0      A:  0%  C:100%  G:  0%  T:  0%       13/26    reads    CC\n",
      "  rs35528968        A:2925   C:0      G:0      T:2      A:100%  C:  0%  G:  0%  T:  0%     2927/5872  reads    AA\n",
      "  rs356167          A:0      C:3358   G:5007   T:0      A:  0%  C: 40%  G: 60%  T:  0%     8365/16859 reads    CG\n",
      "  rs433342          A:5      C:0      G:49     T:0      A:  9%  C:  0%  G: 91%  T:  0%       54/108   reads    GG\n",
      "  rs5030240         A:0      C:3316   G:3267   T:0      A:  0%  C: 50%  G: 50%  T:  0%     6583/13204 reads    CG\n",
      "  rs727241          A:0      C:0      G:0      T:3627   A:  0%  C:  0%  G:  0%  T:100%     3627/7284  reads    TT\n",
      "  rs9274701_alt     A:0      C:4134   G:3      T:0      A:  0%  C:100%  G:  0%  T:  0%     4137/8275  reads    CC\n",
      "  rs9329104         A:1651   C:0      G:0      T:0      A:100%  C:  0%  G:  0%  T:  0%     1651/3510  reads    AA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Get all sample genotypes\n",
    "#\n",
    "excludeSnp = [] #['rs9274701'] # Amplicon is found, but content is not as expected\n",
    "genotypes  = {}\n",
    "\n",
    "for sampleName in sorted(sampleList):\n",
    "  print(\"SNP's for sample {}\".format(sampleName))\n",
    "  genotypes[sampleName] = {}\n",
    "  vcfFile               = os.path.join(mappedDataDir, '{}_direct_mapping.vcf'.format(sampleName))\n",
    "  \n",
    "  d = genotype(vcfFile, 26)\n",
    "  \n",
    "  for snpName in sorted(d):\n",
    "    t = sum(d[snpName]['alleles'].values())\n",
    "    \n",
    "    # Likely genotype\n",
    "    gg = {}\n",
    "    vv = sorted([int(c) for c in d[snpName]['alleles'].values()], reverse=True)\n",
    "    \n",
    "    for g, c in d[snpName]['alleles'].items():\n",
    "      if c not in gg:\n",
    "        gg[c] = [g]\n",
    "      else:\n",
    "        gg[c].append(g)\n",
    "    # Second allele must be at least 10% of main allele or it is ignored.\n",
    "    # Third and fourth alleles are always ignored if present.\n",
    "    #if len(vv) > 1 and vv[1] > vv[0]/10:\n",
    "    # Second allele must be at least 25% of main allele or it is ignored.\n",
    "    if len(vv) > 1 and vv[1] > vv[0]/4:\n",
    "      if len(gg[vv[0]]) > 1:\n",
    "        aa = ''.join(gg[vv[0]])\n",
    "      else:\n",
    "        aa = gg[vv[0]][0] + gg[vv[1]][0]\n",
    "    else:\n",
    "      aa = gg[vv[0]][0] + gg[vv[0]][0]\n",
    "    \n",
    "    o = '  {:<14}  '.format(snpName)\n",
    "    \n",
    "    for g in sorted(d[snpName]['alleles']):\n",
    "      o += '  {}:{:<5}'.format(g, d[snpName]['alleles'][g])\n",
    "    for g in sorted(d[snpName]['alleles']):\n",
    "      if t > 0:\n",
    "        o += '  {}:{:>3}%'.format(g, round(100*d[snpName]['alleles'][g]/t))\n",
    "      else:\n",
    "        o += '  {}:{:>3}%'.format(g, '???')\n",
    "    o += '    {:>5}/{:<5} reads'.format(t, d[snpName]['depth'])\n",
    "    o += '    {}'.format(''.join(sorted(aa)))\n",
    "    genotypes[sampleName][snpName] = ''.join(sorted(aa))\n",
    "    print(o)\n",
    "  \n",
    "  print('')\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
